src/ngraph/runtime/gpu/cublas_emitter.cpp

```diff
@@ -0,0 +1,240 @@
+                                              const Shape& arg0_shape,
+                                              const Shape& arg1_shape,
+                                              const Shape& out_shape,
+                                              size_t reduction_axes)
+{
+    std::stringstream ss;
+    ss << "dot_op"
+       << "_dtype_" << dtype.c_type_string() << "_reduction_axes_count_" << reduction_axes;
+    std::string hash = ss.str() + "_i_" + join(arg0_shape, "_") + "_i_" + join(arg1_shape, "_") +
+                       "_o_" + join(out_shape, "_");
+
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    std::unique_ptr<gpu::primitive> dot;
+    if (arg0_shape.empty() || arg1_shape.empty())
+    {
+        auto& second = (arg0_shape.empty() ? arg1_shape : arg0_shape);
+        int count = shape_size(second);
+
+        size_t firstIndex = (arg0_shape.empty() ? 0 : 1);
+        size_t secondIndex = (arg0_shape.empty() ? 1 : 0);
+
+        dot.reset(new gpu::primitive{[=](void** inputs, void** outputs) {
+            CUBLAS_SAFE_CALL(cublasScopy(*m_ctx->cublas_handle,
+                                         count,
+                                         static_cast<const float*>(inputs[secondIndex]),
+                                         1,
+                                         static_cast<float*>(outputs[0]),
+                                         1));
+            CUBLAS_SAFE_CALL(cublasSscal(*m_ctx->cublas_handle,
+                                         count,
+                                         static_cast<const float*>(inputs[firstIndex]),
+                                         static_cast<float*>(outputs[0]),
+                                         1));
+            debug_sync();
+        }});
+
+        return getPrimitiveIndex(dot, hash);
+    }
+
+    if (shape_size(arg0_shape) == 0 || shape_size(arg1_shape) == 0)
+    {
+        size_t elemSize = shape_size(out_shape) * dtype.size();
+        dot.reset(new gpu::primitive{[=](void** inputs, void** outputs) {
+            runtime::gpu::cuda_memset(outputs[0], 0, elemSize);
+            debug_sync();
+        }});
+
+        return getPrimitiveIndex(dot, hash);
+    }
+    // case that can be treat as dot1d
+    if ((arg0_shape.size() == arg1_shape.size()) && (arg0_shape.size() == reduction_axes))
+    {
+        for (int i = 0; i < arg0_shape.size(); i++)
+        {
+            if (arg0_shape[i] != arg1_shape[i])
+            {
+                throw invalid_argument("arg0 and arg1 shape does not match for dot.");
+            }
+        }
+
+        dot.reset(new gpu::primitive{[=](void** inputs, void** outputs) {
+            CUBLAS_SAFE_CALL(cublasSdot(*m_ctx->cublas_handle,
+                                        shape_size(arg0_shape),
```

Comment:
Calculate the count outside the lambda, as this will compute shape_size at runtime instead of compile time.